### PR TITLE
Fix a connection leak

### PR DESCRIPTION
**Pull request by snjeza**
Fixes https://issues.jboss.org/browse/FUSETOOLS-2634

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

### CODE REVIEW

File: src/main/java/io/takari/aether/wagon/OkHttpWagon.java [hunk 1]
```diff
@@ -78,7 +80,7 @@ public void fillInputData(InputData inputData) throws TransferFailedException, R
         throw new TransferFailedException("Failed to transfer file: " + url + ". Return code is: " + statusCode);
       }
       
-      inputData.setInputStream(response.getInputStream());
+      inputData.setInputStream(new ByteArrayInputStream(response.getBytes()));
```
Comment (fbricon) [inline src/main/java/io/takari/aether/wagon/OkHttpWagon.java]: why is this necessary?

Comment (snjeza) [reply src/main/java/io/takari/aether/wagon/OkHttpWagon.java]: Otherwise WagonTest.testNoHeaders will fail.

File: src/main/java/io/takari/aether/okhttp/OkHttpAetherClient.java [hunk 2]
```diff
@@ -288,6 +288,18 @@ public String getHeader(String name) {
...
       return conn.body().byteStream();
     }
+
+    @Override
+    public byte[] getBytes() throws IOException {
+      return conn.body().bytes();
+    }
+
+	@Override
+	public void close() {
```
Comment (apupier) [inline src/main/java/io/takari/aether/okhttp/OkHttpAetherClient.java]: tabs are used although spaces were used in the project

File: src/main/java/io/takari/aether/connector/AetherRepositoryConnector.java [hunk 3]
```diff
@@ -489,10 +489,15 @@ public void run() {
...
     private boolean resourceExist(String uri) throws IOException {
-      if (aetherClient.head(uri).getStatusCode() == 200) {
-        return true;
+      Response head = aetherClient.head(uri);
+      try {
+        if (head.getStatusCode() == 200) {
+          return true;
+        }
+        return false;
+    } finally {
```
Comment (apupier) [inline src/main/java/io/takari/aether/connector/AetherRepositoryConnector.java]: indentation not correct

File: src/main/java/io/takari/aether/connector/AetherRepositoryConnector.java [hunk 4]
```diff
@@ -489,10 +489,15 @@ public void run() {
     }
 
     private boolean resourceExist(String uri) throws IOException {
-      if (aetherClient.head(uri).getStatusCode() == 200) {
-        return true;
+      Response head = aetherClient.head(uri);
+      try {
+        if (head.getStatusCode() == 200) {
```
Comment (apupier) [inline src/main/java/io/takari/aether/connector/AetherRepositoryConnector.java]: what do you think to simplify it with return head.getStatusCode() == 200; directly?

File: src/main/java/io/takari/aether/connector/AetherRepositoryConnector.java [hunk 5]
```diff
@@ -585,8 +590,9 @@ private FileTransfer resumableGet(String uri, File fileInLocalRepository, Transf
 
         //JVZ: this all needs to be moved up to the client
 
-        Response response;
-
+        Response response = null;
+        Closer closer = null;
+        try {
         if (resumeDownloadInProgress) {
```
Comment (apupier) [inline src/main/java/io/takari/aether/connector/AetherRepositoryConnector.java]: i think that it needs to increase indent here

File: src/main/java/io/takari/aether/wagon/OkHttpWagon.java (hunk 1 above)
Comment (ifedorenko) [reply src/main/java/io/takari/aether/wagon/OkHttpWagon.java]: This is indeed incorrect, `fillInputData` cannot close http response, it must return live input stream to the caller (via `InputData` parameter field) and rely on the caller to close the stream.

File: src/main/java/io/takari/aether/connector/AetherRepositoryConnector.java [hunk 6]
```diff
@@ -736,6 +740,9 @@ public void run() {
         }
       } finally {
         latch.countDown();
+        if (response != null) {
```
Comment (ifedorenko) [inline src/main/java/io/takari/aether/connector/AetherRepositoryConnector.java]: why not use try with resources?

File: src/main/java/io/takari/aether/okhttp/OkHttpAetherClient.java [hunk 7]
```diff
@@ -288,6 +288,13 @@ public String getHeader(String name) {
     public InputStream getInputStream() throws IOException {
       return conn.body().byteStream();
     }
+
+    @Override
+    public void close() {
+      if (conn != null) {
```
Comment (ifedorenko) [inline src/main/java/io/takari/aether/okhttp/OkHttpAetherClient.java]: I don't believe this can be `null` but it is probably a good idea to add a null guard to `ResponseAdapter` constructor

File: src/main/java/io/takari/aether/wagon/OkHttpWagon.java [hunk 8]
```diff
@@ -332,4 +329,59 @@ public void setHttpHeaders(Properties httpHeaders) {
     this.httpHeaders = Collections.unmodifiableMap(map);
   }
 
+  class InputStreamAdapter extends InputStream {
```
Comment (ifedorenko) [inline src/main/java/io/takari/aether/wagon/OkHttpWagon.java]: This isn't necessary, `response.close()`, `response.body().close()` and `response.body().byteStream().close()` all do the same thing.

If you prefer to keep InputStreamAdapter for readability reasons, please subclass `java.io.FilterInputStream`.

File: src/main/java/io/takari/aether/wagon/OkHttpWagon.java [hunk 9]
```diff
@@ -78,7 +78,7 @@ public void fillInputData(InputData inputData) throws TransferFailedException, R
         throw new TransferFailedException("Failed to transfer file: " + url + ". Return code is: " + statusCode);
       }
       
-      inputData.setInputStream(response.getInputStream());
+      inputData.setInputStream(new InputStreamAdapter(response.getInputStream(), response));
```
Comment (ifedorenko) [inline src/main/java/io/takari/aether/wagon/OkHttpWagon.java]: please add a comment that closing returned input stream closes response.

**Comment by ifedorenko** [comment]
Merged. Thank you for the fix and apologies for slow response.
